datacompy/core.py

```diff
@@ -761,3 +754,34 @@ def calculate_max_diff(col_1, col_2):
         return (col_1.astype(float) - col_2.astype(float)).abs().max()
     except:
         return 0
+
+
+def generate_id_within_group(dataframe, join_columns):
+    """Generate an ID column that can be used to deduplicate identical rows.  The series generated
+    is the order within a unique group, and it handles nulls.
+
+    Parameters
+    ----------
+    dataframe : Pandas.DataFrame
+        The dataframe to operate on
+    join_columns : list
+        List of strings which are the join columns
+
+    Returns
+    -------
+    Pandas.Series
+        The ID column that's unique in each group.
+    """
+    default_value = "DATACOMPY_NULL"
```

Comment:
Is the expectation this is passed in with the DF? I'm probably missing the context around this, but where does `DATACOMPY_NULL` get set or come from?

Reply:
This is to handle cases where there are both nulls and duplicates in the columns you're joining on.

I was using `df.groupby(join_columns).cumcount()` to generate an ID that was unique within each unique group, but because of how groupby drops nulls, anything with a null in a join_column would just get lumped into one big group.

To handle that case we need to `fillna` before `groupby`.  In this PR, if there are nulls in the join columns, I convert all the join columns to text, then just fill nulls with `DATACOMPY_NULL`, which (hopefully) shouldn't exist in your dataframe.  If it does exist it'll fail, and the user will have to fillna themselves before comparing.

Guess I should probably add something in the docs too :)

Reply:
Makes sense!
Yeah a blurb in the docs might be helpful for the context around this.